packages/api/src/controllers/stripe.ts

```diff
@@ -0,0 +1,720 @@
+import Router from "express/lib/router";
+import { db } from "../store";
+import Stripe from "stripe";
+import { products } from "../config";
+import sql from "sql-template-strings";
+import fetch from "node-fetch";
+import qs from "qs";
+import { sendgridEmailPaymentFailed } from "./helpers";
+
+const app = Router();
+
+interface OverUsageBill {
+  transcodingBill: OverUsageItem;
+  deliveryBill: OverUsageItem;
+  storageBill: OverUsageItem;
+}
+
+interface OverUsageItem {
+  units: number;
+  total: number;
+}
+
+interface BillingUsageData {
+  DeliveryUsageMins: number;
+  TotalUsageMins: number;
+  StorageUsageMins: number;
+}
+
+const testProducts = ["hacker_1", "growth_1", "scale_1"];
+
+const productMapping = {
+  hacker_1: "prod_O9XuIjn7EqYRVW",
+  growth_1: "prod_O9XtHhI6rbTT1B",
+  scale_1: "prod_O9XtcfOSMjSD5L",
+};
+
+export const reportUsage = async (req, adminToken) => {
+  const [users] = await db.user.find(
+    [
+      sql`users.data->>'stripeProductId' IN ('growth_1', 'scale_1', 'prod_O9XtHhI6rbTT1B','prod_O9XtcfOSMjSD5L')`,
+    ],
+    {
+      limit: 9999999999,
+      useReplica: true,
+    }
+  );
+
+  const [oldProPlanUsers] = await db.user.find(
+    [
+      sql`users.data->>'oldProPlan' = true AND users.data->>'stripeProductId' IN ('hacker_1','prod_O9XuIjn7EqYRVW')`,
+    ],
+    {
+      limit: 9999999999,
+      useReplica: true,
+    }
+  );
+
+  // Join oldProPlanUsers and users
+  const payAsYouGoUsers = [...users, ...oldProPlanUsers];
+
+  let updatedUsers = [];
+  for (const user of payAsYouGoUsers) {
+    const userSubscription = await req.stripe.subscriptions.retrieve(
+      user.stripeCustomerSubscriptionId
+    );
+
+    const billingCycleStart = userSubscription.current_period_start * 1000; // 1685311200000 // Test date
+    const billingCycleEnd = userSubscription.current_period_end * 1000; // 1687989600000 // Test date
+
+    const ingests = await req.getIngest();
+    const billingUsage = await getBillingUsage(
+      user.id,
+      billingCycleStart,
+      billingCycleEnd,
+      ingests[0].origin,
+      adminToken
+    );
+
+    const overUsage = await calculateOverUsage(
+      products[user.stripeProductId],
+      billingUsage
+    );
+
+    const subscriptionItems = await req.stripe.subscriptionItems.list({
+      subscription: user.stripeCustomerSubscriptionId,
+    });
+
+    // create a map of subscription items by their lookup keys
+    const subscriptionItemsByLookupKey = subscriptionItems.data.reduce(
+      (acc, item) => {
+        acc[item.price.lookup_key] = item.id;
+        return acc;
+      },
+      {} as Record<string, string>
+    );
+
+    try {
```

Comment:
Would be nice to wrap the whole iteration with the `try/catch` to make sure we don't fail-fast on errors in the above calls. Ideally could also move the `try` body to a helper func so we can have a clearer `try/call/catch/updateResult` logic here